lib/Dialect/TritonGPU/Transforms/LoopScheduling.cpp

```diff
@@ -0,0 +1,198 @@
+
+      for (auto op : backwardSlice) {
+        if (auto ifOp = dyn_cast<scf::IfOp>(op)) {
+          ifsToStage.insert({ifOp, stage});
+        }
+      }
+    }
+  }
+  tt::CoarseSchedule::Cluster prologueCluster = schedule.clusters.newAtFront();
+  for (auto [ifOp, stage] : ifsToStage) {
+    schedule.insert(ifOp, stage, prologueCluster);
+  }
+  // Look for the IfOp that is in the forward slice of the root users and put it
+  // at the end of the loop.
+  tt::CoarseSchedule::Cluster epilogueCluster = schedule.clusters.newAtBack();
+  for (auto rootUser : rootUsers) {
+    SetVector<Operation *> forwardSlice;
+    getForwardSlice(rootUser, &forwardSlice);
+
+    int stage = schedule[rootUser].first;
+    for (auto op : forwardSlice) {
+      scf::IfOp ifOp = dyn_cast<scf::IfOp>(op);
+      if (ifOp == nullptr) {
+        // check if the op is in the body of an if op that's part of the loop
+        auto parentOp = op->getParentOp();
+        if (parentOp != nullptr &&
+            parentOp->getParentOp() == forOp.getOperation()) {
+          ifOp = dyn_cast<scf::IfOp>(parentOp);
+        }
+      }
+      if (ifOp) {
+        schedule.insertIfAbsent(ifOp, stage,
+                                epilogueCluster); // after prefetch extracts
+      }
+    }
+  }
+  return afterPrologue;
+}
+
+#define GEN_PASS_DEF_TRITONGPULOOPSCHEDULING
+#include "triton/Dialect/TritonGPU/Transforms/Passes.h.inc"
+
+class TritonGPULoopSchedulingPass
+    : public impl::TritonGPULoopSchedulingBase<TritonGPULoopSchedulingPass> {
+public:
+  using impl::TritonGPULoopSchedulingBase<
+      TritonGPULoopSchedulingPass>::TritonGPULoopSchedulingBase;
+
+  int getNumStagesOrDefault(scf::ForOp forOp) {
+    // Use the attribute attached to the loop if it exists otherwise use the
+    // global control.
+    if (!forOp->hasAttr(mlir::triton::kNumStagesAttrName))
+      return numStages;
+    return mlir::cast<IntegerAttr>(
+               forOp->getAttr(mlir::triton::kNumStagesAttrName))
+        .getInt();
+  }
+
+  void runOnOperation() override {
+    SmallVector<scf::ForOp> loops;
+    getOperation()->walk([&](scf::ForOp forOp) {
+      // Bail out for loops with num_stage <= 1.
+      if (getNumStagesOrDefault(forOp) > 1)
+        loops.push_back(forOp);
+    });
+
+    if (loops.empty())
+      return;
+    for (scf::ForOp forOp : loops) {
+      int loopNumStages = getNumStagesOrDefault(forOp);
+      DenseSet<Operation *> rootUsers;
+      tt::CoarseSchedule coarseSchedule(loopNumStages);
+      scheduleLoads(forOp, coarseSchedule, rootUsers, loopNumStages);
+      tt::CoarseSchedule::Cluster afterPrologue = schedulePrologueAndEpilogue(
+          forOp, coarseSchedule, rootUsers, loopNumStages);
+      // Go through schedule and assign (stage, cluster).
+      // shift so afterPrologue will be at clusterId 0
+      auto ctx = forOp.getContext();
+      for (auto [op, stage_, cluster] : coarseSchedule.getOpsInOrder(forOp)) {
+        op->setAttr("loop.stage",
```

Comment:
Will the two new attributes be removed after the pipelining pass?

Reply:
Do we want to do the stage assignments for inner-most loop only? Wondering how to avoid overwriting stages in the presence of multi-dimensional loops.

Reply:
Does it make sense to have the attribute name store somewhere in utilities so that it can be shared by the two passes?